### PR TITLE
Fix typed array export... again

### LINKED ISSUE #62916: Can't assign nodes to exported array
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

Godot 4.0 alpha 11

### System information

Solus OS and Windows

### Issue description

Can't assign nodes to an exported array of node pointers

https://user-images.githubusercontent.com/61121689/178321358-fe05eec9-17ab-4d94-ac87-f87b1e059b97.mp4

### Steps to reproduce

1) Export an array of node pointers (`@export var node_pointer_array: Array[Node2D]`)
2) Try to assign an element to the array on the editor
3) Element will not be assigned

Or: check out the minimal reproduction project below

### Minimal reproduction project

[export array of node pointers.zip](https://github.com/godotengine/godot/files/9086249/export.array.of.node.pointers.zip)

### PR DESCRIPTION
**Pull request by KoBeWi**
Fixes #62916 again (regression from #76114)

CC @ajreckof

### CODE REVIEW

File: editor/editor_properties_array_dict.cpp [hunk 1]
```diff
@@ -203,7 +203,18 @@ void EditorPropertyArray::_property_changed(const String &p_property, Variant p_
 		index = p_property.get_slice("/", 1).to_int();
 	}
 
-	Variant array = object->get_array().duplicate();
+	Variant array;
+	const Variant &original_array = object->get_array();
+
+	if (original_array.get_type() == Variant::ARRAY) {
+		// Needed to preserve type of TypedArrays in meta pointer properties.
```
Comment (KoBeWi) [inline editor/editor_properties_array_dict.cpp:210]: tbh I don't know when exactly it's needed, but it's specifically to fix Node arrays.

**Comment by ajreckof** [comment]
I am so sorry, I didn't know Node-typed Arrays where acting different so I only tested with one typed array and figured every other one would work the same. If i understand well what is happening here for node-typed exports `_property_changed` do not return a `Node` but only its `NodePath` (which i feel is strange). So you need to untype the array to be able to set and then retype the array later and at the same type convert the `NodePath` to a `Node`. If Nodes are the only strange case maybe we should detect if the value is a `NodePath` and if the subtype of the array is Object and in this case directly convert the `NodePath` to a `Node`.  

I've given a test to your fix and it seems that it breaks if you try to use a property with a setter. What is happening is that because of the mismatch in type the setter is never called (but no error which is strange).

So I went a bit further and discovered that it is also a problem with direct Node export. I feel it would be better to try and fix the Node export so that they effectively emit property changed with a Node and not a NodePath.  

https://user-images.githubusercontent.com/66184050/233867421-6480fc35-6ec2-4a60-aeda-5f2437169c8d.mp4

resolving the problem shown just above would also resolve the current problem.

**Comment by KoBeWi** [comment]
I think this PR could be merged for 4.0.3, as it's less risky than your new solution.
Or maybe we could only cherry-pick the original fix and live with #76113 (as it's a bit corner-case).

**Comment by ajreckof** [comment]
As bothe the original issue and the issue found above are two corner cases, so i feel it won't be a problem if either of them is kept for a few more iterations until my solution is polished and thoroughly tested.
Whether my PR being undone or your fix being applied i feel it would be better to have your PR would be a bit better and it is already done so why take advantage of it.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
We have discussed this with the production team and the overall opinion is that cherry-picking it would risk the stability of 4.0. There are follow-up fixes some of which haven't been merged or battle-tested yet, and the more of them we have to apply, the bigger the scope and the risk gets. So I'm removing the label for that reason. Please update to 4.1 if this affects your projects.

**Pull request merged**